modules/caddyhttp/httpredirectlistener.go

```diff
@@ -86,39 +86,54 @@ func (l *httpRedirectListener) Accept() (net.Conn, error) {
 	}
 
 	return &httpRedirectConn{
-		Conn: c,
-		r:    bufio.NewReader(io.LimitReader(c, maxHeaderBytes)),
+		Conn:  c,
+		limit: maxHeaderBytes,
+		r:     bufio.NewReader(c),
 	}, nil
 }
 
 type httpRedirectConn struct {
 	net.Conn
-	once sync.Once
-	r    *bufio.Reader
+	once  bool
+	limit int64
+	r     *bufio.Reader
 }
 
 // Read tries to peek at the first few bytes of the request, and if we get
 // an error reading the headers, and that error was due to the bytes looking
 // like an HTTP request, then we perform a HTTP->HTTPS redirect on the same
 // port as the original connection.
 func (c *httpRedirectConn) Read(p []byte) (int, error) {
-	var errReturn error
-	c.once.Do(func() {
+	// no need to use sync.Once - net.Conn is not read from concurrently.
+	if !c.once {
+		c.once = true
+
 		firstBytes, err := c.r.Peek(5)
 		if err != nil {
-			return
+			return 0, err
 		}
 
 		// If the request doesn't look like HTTP, then it's probably
-		// TLS bytes and we don't need to do anything.
+		// TLS bytes, and we don't need to do anything.
+		// 0, nil are valid return values - caller most likely will try again.
 		if !firstBytesLookLikeHTTP(firstBytes) {
-			return
+			return 0, nil
```

Comment:
Would it be better to `return c.r.Read(p)` here?

Reply:
I thought about that, but decided to keep it as it was out of habit.

Reply:
Would TLS perform slightly worse by having to re-enter on the first read? Probably nanoseconds at worst but :man_shrugging:

Reply:
☝️ @WeidiDeng I'm good to merge once this is answered. I don't think it's a big deal, but just a thought, since this is definitely a hot path when enabled so any work that can be avoided is a good thing.